src/t_hash.c

```diff
@@ -558,6 +875,183 @@ void hashTypeRandomElement(robj *hashobj, unsigned long hashsize, listpackEntry
+ */
+static ExpireAction hashTypeActiveExpire(eItem _hashObj, void *dbCtx) {
+    robj *hashObj = (robj *) _hashObj;
+    redisDb *db = (redisDb *) dbCtx;
+
+    if (hashObj->encoding == OBJ_ENCODING_LISTPACK) {
+        serverPanic("Listpack encoding not supported yet");
+    }
+    serverAssert(hashObj->encoding == OBJ_ENCODING_HT);
+
+    dict *d = hashObj->ptr;
+    dictExpireMetadata *dictExpireMeta = (dictExpireMetadata *) dictMetadata(d);
+
+    ExpireInfo info = {
+            .maxToExpire = 0xFFFF,
+            .onExpireItem = onFieldExpire,
+            .ctx = hashObj,
+            .now = commandTimeSnapshot(),
+            .itemsExpired = 0};
+
+    ebExpire(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, &info);
+
+    /* If hash has no more fields to expire, remove it from HFE DB */
+    if (info.nextExpireTime == 0) {
+        if (hashTypeLength(hashObj, 0) == 0) {
+            robj *key = createStringObject(dictExpireMeta->key, sdslen(dictExpireMeta->key));
+            dbDelete(db, key);
+            //notifyKeyspaceEvent(NOTIFY_HASH,"xxxxxxxxx",c->argv[1],c->db->id);
+            notifyKeyspaceEvent(NOTIFY_GENERIC,"del",key, db->id);
+            server.dirty++;
+            signalModifiedKey(NULL, &server.db[0], key);
+            decrRefCount(key);
+        }
+        return ACT_REMOVE_EXP_ITEM;
+    } else {
+        /* Hash has more fields to expire. Keep hash to pending items that will
+         * be added back to global HFE DS at the end of ebExpire() */
+        ExpireMeta *expireMeta = hashGetExpireMeta(hashObj);
+        ebSetMetaExpTime(expireMeta, info.nextExpireTime);
+        return ACT_UPDATE_EXP_ITEM;
+    }
+}
+
+/* Delete all expired fields in a hash */
+static void hashTypeDeleteExpiredFields(client *c, robj *hashObj) {
+    redisDb *db = c->db;
+
+    if (hashObj->encoding == OBJ_ENCODING_LISTPACK) {
+        return; /* TODO */
+    }
+    serverAssert(hashObj->encoding == OBJ_ENCODING_HT);
+
+    dict *d = hashObj->ptr;
+
+    if (!isDictWithMetaHFE(d))
+        return;
+
+    dictExpireMetadata *dictExpireMeta = (dictExpireMetadata *) dictMetadata(d);
+
+    /* If HFE metadata is marked as trash, return */
+    if (dictExpireMeta->expireMeta.trash)
+        return;
+
+    /* If time of next hash-field to expire is greater than current time, return */
+    if (ebGetExpireTime(&hashExpireBucketsType, hashObj) >= (uint64_t)commandTimeSnapshot() )
+        return;
+
+    /* Remove expired fields as part of lazy-expire */
+    ExpireInfo info = {
+            .maxToExpire = 0xFFFFFFFF,
+            .onExpireItem = onFieldExpire,
+            .ctx = hashObj,
+            .now = commandTimeSnapshot(),
+            .itemsExpired = 0};
+
+    ebExpire(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, &info);
+
+    /* If hash has no more fields to expire, remove it from HFE DB */
+    if (info.nextExpireTime == 0) {
+
+        if (hashTypeLength(hashObj, 0) == 0) {
```

Comment:
IIUC this is not going to delete the hash from the global "db->hexpires" if there are fields with no TTL.

Reply:
@tezc nice catch. 
If there are no more HFEs but the hash is still not empty with persist fields then I forget to remove it from `db->hexpires`. 

Thanks!